builtin/aws/ssm/config_sourcer.go

```diff
@@ -0,0 +1,383 @@
+			result.Result = &pb.ConfigSource_Value_Error{
+				Error: status.New(codes.Aborted, err.Error()).Proto(),
+			}
+
+			continue
+		}
+
+		// Store our request
+		requests[req.Name] = &decodedReq
+	}
+
+	// Start the refresher. Note for the log parameter we can't use "log"
+	// because that is only alive for the duration of the plugin RPC call.
+	// Once the RPC completes we'll get errors. So we instantiate a new
+	// logger here which will go to the plugin process stderr.
+	refreshCtx, cancel := context.WithCancel(context.Background())
+	cs.refreshCancel = cancel
+	initCh := make(chan struct{})
+	go cs.startRefresher(refreshCtx,
+		hclog.L().Named("ssm-refresher"),
+		initCh, requests, cs.values, ssmsvc)
+
+	// Unlock so that the refresher can populate.
+	log.Debug("waiting for first set of values")
+	cs.cacheMu.Unlock()
+	<-initCh
+	cs.cacheMu.Lock()
+
+	// Return our results
+	log.Debug("first set of values received, returning")
+	result := make([]*pb.ConfigSource_Value, 0, len(cs.values))
+	for _, v := range cs.values {
+		result = append(result, v)
+	}
+
+	return result, nil
+}
+
+func (cs *ConfigSourcer) stop() error {
+	cs.cacheMu.Lock()
+	defer cs.cacheMu.Unlock()
+
+	// Nullify everything which will force a refresh and restart
+	cs.values = nil
+
+	return nil
+}
+
+func (cs *ConfigSourcer) startRefresher(
+	ctx context.Context,
+	log hclog.Logger,
+	initCh chan<- struct{},
+	wpreqs map[string]*reqConfig,
+	values map[string]*pb.ConfigSource_Value,
+	client *ssm.SSM,
+) {
+	// We need to keep a map of our parameter names to the
+	// value keys that require it.
+	paramToReq := map[string][]string{}
+
+	// Build our actual request parameter. This remains constant so
+	// we calculate this once.
+	req := &ssm.GetParametersInput{}
+	req.SetWithDecryption(true)
+	for k, wpreq := range wpreqs {
+		name := wpreq.Path
+		req.Names = append(req.Names, &name)
+		paramToReq[name] = append(paramToReq[name], k)
+	}
+
+	// Calculate a sleep period with a 30% jitter added to it.
+	const factor = 0.5
+	min := int64(math.Floor(float64(refreshPeriod) * (1 - factor)))
+	max := int64(math.Ceil(float64(refreshPeriod) * (1 + factor)))
+
+	for {
+		// Read our value
+		log.Trace("querying parameters")
+		resp, err := client.GetParameters(req)
+		if err != nil {
+			// Just skip these errors and reuse the last known values for
+			// all requests that we have.
+			log.Warn("error querying parameters", "err", err)
+			continue
+		}
+		log.Trace("parameters received", "len", len(resp.Parameters))
+
+		// Update our values
+		cs.cacheMu.Lock()
+		for _, param := range resp.Parameters {
+			if param.Name == nil || param.Value == nil {
+				// should never happen
+				log.Warn("param name or value is nil", "param", param)
+				continue
+			}
+
+			for _, k := range paramToReq[*param.Name] {
+				v, ok := values[k]
+				if !ok {
+					// skip params we don't know, since we should prepopulate it all.
```

Comment:
Could you expand on this comment? I'm not sure why we're updating known values but ignoring new ones.

Reply:
Yes, the short of it is: this should never happen. The why is: we prepopulate `values` and requests on/around line 97. If AWS is well behaved, it should only send us the things we requested. I assume that's certainly true, but we have this check here to avoid a nil exception in case it isn't true.

Reply:
AH HA! It was a question of _what_ was prepopulated that I missed.